spring-web/src/main/java/org/springframework/web/service/invoker/PathVariableArgumentResolver.java

```diff
@@ -0,0 +1,140 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.web.service.invoker;
+
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.core.MethodParameter;
+import org.springframework.core.ReactiveAdapter;
+import org.springframework.core.ReactiveAdapterRegistry;
+import org.springframework.core.convert.ConversionService;
+import org.springframework.core.convert.TypeDescriptor;
+import org.springframework.lang.Nullable;
+import org.springframework.util.StringUtils;
+import org.springframework.web.bind.annotation.PathVariable;
+
+/**
+ * An implementation of {@link HttpServiceMethodArgumentResolver} that resolves request path variables
+ * based on method arguments annotated with {@link  PathVariable}. {@code null} values are allowed only if {@link PathVariable#required()}
+ * is {@code true}.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 6.0
+ */
+public class PathVariableArgumentResolver implements HttpServiceMethodArgumentResolver {
+
+	private static final Log LOG = LogFactory.getLog(PathVariableArgumentResolver.class);
+	private static final TypeDescriptor STRING_TYPE_DESCRIPTOR = TypeDescriptor.valueOf(String.class);
+	private final ConversionService conversionService;
+
+	public PathVariableArgumentResolver(ConversionService conversionService) {
+		this.conversionService = conversionService;
+	}
+
+	// TODO: handle Maps without name in annotation
+	@Override
+	public void resolve(@Nullable Object argument, MethodParameter parameter, HttpRequestDefinition requestDefinition) {
+		if (!parameter.hasParameterAnnotation(PathVariable.class)) {
+			return;
+		}
+		PathVariable annotation = parameter.getParameterAnnotation(PathVariable.class);
+		String resolvedAnnotationName = StringUtils.hasText(annotation.value()) ? annotation.value() : annotation.name();
+		Object resolvedArgument = resolveFromOptional(argument);
+		Class<?> type = parameter.nestedIfOptional().getNestedParameterType();
+		boolean required = annotation.required();
+		validateForReactiveWrapper(type);
+		validateForNullArgument(resolvedArgument, required);
+		if (Map.class.isAssignableFrom(type) && StringUtils.hasText(resolvedAnnotationName)) {
+			if (resolvedArgument instanceof Map<?, ?>) {
+				Map<?, ?> valueMap = (Map<?, ?>) resolvedArgument;
+				addNamedPathVariableFromMap(requestDefinition, valueMap, resolvedAnnotationName, required);
+				return;
+			}
+		}
+		String name = StringUtils.hasText(resolvedAnnotationName) ? resolvedAnnotationName : parameter.getParameterName();
+		String stringValue = convertToString(new TypeDescriptor(parameter.nestedIfOptional()), resolvedArgument);
+		if (LOG.isTraceEnabled()) {
+			LOG.trace("Path variable " + resolvedAnnotationName + " resolved to " + stringValue);
+		}
+		requestDefinition.getUriVariables().put(name, stringValue);
```

Comment:
There is some overlap in adding a variable value, whether it's from a Map or not, including validation, type conversion, and logging). Maybe this could be encapsulated a little better.